libs/vscode-ui/components/src/lib/autocomplete/autocomplete.component.ts

```diff
@@ -1,16 +1,222 @@
+              switchMap(visibleOptions =>
+                visibleOptions.length
+                  ? fromEvent<KeyboardEvent>(document, 'keyup').pipe(
+                      map((event: KeyboardEvent) => event.key),
+                      filter(key => isNavigationKey(key)),
+                      scan(
+                        ([index, _]: [number, string | null], key) =>
+                          [
+                            updatedOptionIndex(
+                              <AutocompleteNavKeys>key,
+                              index || 0,
+                              visibleOptions.length
+                            ),
+                            key
+                          ] as [number, string | null],
+                        [0, null] as [number, string | null]
+                      ),
+                      startWith([0, null] as [number, string | null]),
+
+                      tap(([index, key]) => {
+                        if (key === AutocompleteNavKeys.Enter) {
+                          this.optionSelected(visibleOptions[index]);
+                        }
+                      }),
+
+                      map(([selectedIndex]) => selectedIndex)
+                    )
+                  : of(null)
+              )
+            )
+          : of(null)
+      ),
+      shareReplay(1)
+    );
+
+    this.selectedOption$.subscribe();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes.options) {
+      this._options$.next(this.options);
+    }
+  }
+
+  optionSelected(value: string) {
+    this.control.setValue(value);
+    this.textInput.nativeElement.blur();
+  }
+
+  registerOnChange(fn: any) {
+    this.control.valueChanges
+      .pipe(
+        tap(fn),
+        takeUntil(this.destroying)
+      )
+      .subscribe();
+  }
+
+  registerOnTouched() {}
+
+  setDisabledState(isDisabled: boolean) {
+    isDisabled ? this.control.disable() : this.control.enable();
+  }
+
+  ngOnDestroy() {
+    this.destroying.next();
+  }
+}
+
+/**
+ * Check if the input key is one relevant to autocomplete panel navigation
+ */
+function isNavigationKey(key: KeyboardEvent['key']): boolean {
+  return (AUTOCOMPLETE_NAV_KEYS as string[]).includes(key);
+}
+
+/**
+ * Updates the index of a selected option based on user input. If the new index
+ * is beyond the scope of the available indices, it should return an index on other
+ * side of the array (to loop around the list).
+ * @param key {AutocompleteNavKeys} KeyboardEvent.key value
+ * @param currentIndex {number} index before user input
+ * @param max {number} total number of options
+ * @returns new option index
+ */
+function updatedOptionIndex(
+  key: AutocompleteNavKeys,
+  currentIndex: number,
+  max: number
+): number {
+  let modifier: -1 | 0 | 1 = 0;
+
+  if (key === AutocompleteNavKeys.ArrowDown) {
+    modifier++;
+  }
+  if (key === AutocompleteNavKeys.ArrowUp) {
+    modifier--;
+  }
+
+  const temp = currentIndex + modifier;
```

Comment:
nit: consider calling this something like `selectedIndex` for code clarity.

Reply:
done